src/tiramisu_core.cpp

```diff
@@ -7810,4 +7810,231 @@ void tiramisu::buffer::tag_gpu_register() {
     set_auto_allocate(false);
 }
 
+
+/**
+  * This function generates communication code for data exchange between ranks.
+  * Not safe to. currently being implemented
+  * Details:
+  * Suppose we have the following code
+  * \code
+  * data.set_expression(in(i)+in(i-1));
+  * data.tag_distribute_level(i);
+  * data.generate_communication(number_of_ranks)
+  * /endcode
+  * We have a computation data and an input in, both data and input have i as
+  * their distributed dimension, number_of_ranks is the number of mpi ranks.
+  * this function will generate send/receive to exchange all data "data" required by a
+  * rank and not owned by it.
+ */
+void computation::generate_communication(int number_of_nodes){
+    DEBUG_FCT_NAME(10);
+    DEBUG_INDENT(4);
+
+    //get the distribution map of this rank r
+    isl_map* distribution_map_r=construct_distribution_map("r",number_of_nodes);
+    DEBUG(3, tiramisu::str_dump("The distribution_map for this rank : ");
+        isl_map_dump(distribution_map_r));
+
+    //get the distribution map for other ranks rp
+    isl_map* distribution_map_rp=construct_distribution_map("rp",number_of_nodes);
+    DEBUG(3, tiramisu::str_dump("The distribution_map for other rank : ");
+        isl_map_dump(distribution_map_rp));
+
+    //get the initial domain of the computation
+    isl_set* initial_domain=isl_set_copy(this->get_iteration_domain());
+    DEBUG(3, tiramisu::str_dump("Initial dommain of the computation : ");
+        isl_set_dump(initial_domain));
+
+    //get the set that will be computed by the rank r for this computation
+    isl_set* to_compute_set=isl_set_apply(isl_set_copy(initial_domain),isl_map_copy(distribution_map_r));
+    DEBUG(3, tiramisu::str_dump("Initial dommain of the computation : ");
+        isl_set_dump(to_compute_set));
+
+    //To compute the data needed by this computation,
+    //we have to retrieve its read accesses.
+    std::vector<isl_map*> read_accesses;
+    generator::get_rhs_accesses(this->get_function(),this,read_accesses,false);
+    std::vector<isl_set*> needed_sets;
+    //apply the map to get the set of the needed elements due to a read
+
+    if(read_accesses.size()==0) return; //no read accesses - for inputs
+
+    //calculate the needed sets
+    for(int i=0;i<read_accesses.size();i++){
+        needed_sets.push_back(isl_set_apply(isl_set_copy(to_compute_set),read_accesses[i]));
+    }
+
+    //construct a union set of the needed sets
+    isl_union_set* need_set;
+    if(needed_sets.size()>0) need_set=isl_union_set_from_set(isl_set_copy(needed_sets[0]));
+    for(int i=1;i<needed_sets.size();i++){
+         need_set=isl_union_set_union(isl_union_set_copy(need_set),
```

Comment:
No need for "isl_union_set_copy" here. need_set will be freed and a new need_set will be created, so no need for copying it.